### PR TITLE
backupccl: prevent restoring a newer backup

### PR DESCRIPTION
**Pull request by dt**
This adds a check during RESTORE that the backup(s) being restored were produced by a version
no newer than the current version. The build info for the version that produced a
backup is persisted in the manifest and is used for this check.

Note: this is checking the _build_ version, not the cluster version.  A follow-up
change should likely be added to persist cluster versions and check them as well
so that most-migration data cannot be restored to a pre-migration cluster even if
both are on the same binary version.

Release note (ops change): RESTORE now refuses to restore BACKUPs made by newer versions.

### CODE REVIEW

**Comment by pbardea** [comment]
I have a couple of questions:
1. Since patch releases are backwards compatible, I would expect that 20.2.4 backups could be restored on 20.2.1 nodes, correct?
2. When considering mixed-version clusters are we okay with disallowing a restore of a backup made by a node with a later build number? (E.g. mixed version 20.2, 21.1 cluster - a backup taken by a 20.2 node should be restorable by the 21.1 node given the cluster version hasn't been upgraded.) I don't think we currently have any issues allowing such a restore from happening (although, granted is not something super well tested), and it is before the "burn the boats" cluster upgrade finalization step, so it may be a surprise to users.

I think this is a good indication that we may actually need to store the cluster version to do this check? And I think it would be good to start persisting that in 21.1 before it sets sail.

**Comment by dt** [comment]
re:1) ah, yeah, good point, I should just be comparing Major. Done.

2) Yeah, I have another branch with the cluster-version persisted since, as I mention in the commit here, I think we also need that, since even within the same binary running the whole time we could lose compatibility once you run a one-way migrate that should block restoring to pre-migrate version. But I wanted to do that separately given existing backups don't persist it. I'm less sure that I like the strictness that a backups taken after the binary upgrade but before the logical upgrade will only be restorable by the new binary -- that kinda feels like it defeats the point of the soft-upgrade window.

**Comment by pbardea** [comment]
> I'm less sure that I like the strictness that a backups taken after the binary upgrade but before the logical upgrade will only be restorable by the new binary -- that kinda feels like it defeats the point of the soft-upgrade window.

Agreed, and just to spell it out completely, I think that this means that we can't do this build version check in that case right? Since in the mixed-version situation I described a new node may take a backup with build tag 21.1, but that backup would not be restorable by the older nodes in the same cluster that would have build version 20.2.

**Comment by dt** [comment]
Thought about this a little more and opted to just do the cluster version one now instead. I'd initially held off on it, thinking we'd wait to use it instead of build info in a later version once it'd been being persisted for a release, but on reflection it doesn't matter: if only new nodes are doing the check, it is fine if they look for a field only written by nodes at least that new since they're specifically look for _newer_ nodes anyway. Reworked to be all cluster-version.

File: pkg/ccl/backupccl/backup_test.go [hunk 1]
```diff
@@ -7326,6 +7326,55 @@ func TestCleanupDoesNotDeleteParentsWithChildObjects(t *testing.T) {
...
+	// Also write the checksum file to match the new manifest.
+	checksum, err := getChecksum(manifestData)
+	require.NoError(t, err)
+	require.NoError(t, ioutil.WriteFile(manifestPath+backupManifestChecksumSuffix, checksum, 0644 /* perm */))
+
+	// Verify we reject it.
+	sqlDB.ExpectErr(t, "backup from version 99.1 is newer than current version", `RESTORE DATABASE r1 FROM 'nodelocal://0/too_new'`)
+
+	// Bump the version down and write it back out to make it look older.
+	backupManifest.ClusterVersion = roachpb.Version{Major: 20, Minor: 2, Internal: 2}
```
Comment (pbardea) [inline pkg/ccl/backupccl/backup_test.go:7364]: Can we also add a test case where we nil out the cluster version to simulate earlier backups?

File: pkg/ccl/backupccl/restore_planning.go [hunk 2]
```diff
@@ -1584,6 +1584,15 @@ func doRestorePlan(
 		return err
 	}
 
+	currentVersion := p.ExecCfg().Settings.Version.ActiveVersion(ctx)
```
Comment (pbardea) [inline pkg/ccl/backupccl/restore_planning.go:1587]: Perhaps a comment here mentioning that cluster versions are not expected to ever decrease even in the case of rolling back to an earlier patch version would be useful.

**Review by pbardea — APPROVED** [review]
LGTM with minor testing comment

**Review by Elliebababa — APPROVED** [review]

File: pkg/ccl/backupccl/backup_test.go (hunk 1 above)
Comment (dt) [reply pkg/ccl/backupccl/backup_test.go:7364]: good idea. Done.

**Review by pbardea — APPROVED** [review]

**Comment by dt** [comment]
TFTR!

bors r+

**Pull request merged**